utils/data.cpp

```diff
@@ -0,0 +1,253 @@
+            file.close();
+
+            // Update the main canvas once all shapes have been added to it
+            mainCanvas->repaint();
+        }// end try
+        catch (GeneralException& exception){}
+    }// end if (directory is non-empty)
+}
+
+void Data::GetTo(QTextStream &qin, QChar target)
+{
+    QChar buffer;
+
+    // Read in characters into the buffer until the buffer character is a whitespace
+    do
+    {
+        qin >> buffer;
+    }while(buffer != target);
+}
+
+void Data::ExtractFont(QTextStream& qin, QFont& font)
+{
+    qint32 tempInt; // Temporary storage for an integer
+    QString tempString; // Temporary storage for a string
+
+    // Get text size
+    GetTo(qin, TITLE_INTERRUPTOR);
+    qin >> tempInt;
+    font.setPointSize(tempInt);
+
+    // Get font family
+    GetTo(qin, TITLE_INTERRUPTOR);
+    tempString = qin.readLine();
+    font.setFamily(tempString);
+
+    // Get font style
+    GetTo(qin, TITLE_INTERRUPTOR);
+    qin >> tempString;
+    font.setStyle(StringToEnum(tempString, TEXT_STYLE_STR, TEXT_STYLE_VAL));
+
+    // Get font weight
+    GetTo(qin, TITLE_INTERRUPTOR);
+    qin >> tempString;
+    font.setWeight(StringToEnum(tempString, TEXT_WEIGHT_STR, TEXT_WEIGHT_VAL));
+}
+
+Shape* Data::ConstructLine(QString& type, QList<qint32>& dimensions)
+{
+    if(type == LINE)
+    {
+       return new Line(QPoint(dimensions[0], dimensions[1]), QPoint(dimensions[2], dimensions[3]));
+    }
+    else if(type == POLYLINE)
+    {
+        return new Polyline(GetPoints(dimensions), uint32_t(dimensions.size() / 2));
+    }
+    else
+    {
+        QString message;
+        QTextStream stream(&message);
+        stream << type << " cannot be meaningfully interpreted as a shape type by ConstructLine method";
+        GeneralException(message).raise();
+        return nullptr;
+    }
+}
+
+Shape* Data::ConstructShape(QString& type, QList<qint32>& dimensions)
+{
+    if(type == POLYGON)
+    {
+        return new Polygon(GetPoints(dimensions), uint32_t(dimensions.size() / 2));
+    }
+    else if(type == RECTANGLE)
+    {
+        return new Rect(QPoint(dimensions[0], dimensions[1]), dimensions[2], dimensions[3]);
+    }
+    else if(type == SQUARE)
+    {
+        return new Rect(QPoint(dimensions[0], dimensions[1]), dimensions[2], dimensions[2]);
+    }
+    else if(type == ELLIPSE)
+    {
+        return new Ellipse(QPoint(dimensions[0], dimensions[1]), dimensions[2], dimensions[3]);
+    }
+    else if(type == CIRCLE)
+    {
+        return new Circle(QPoint(dimensions[0], dimensions[1]), dimensions[2]);
+    }
+    else
+    {
+        QString message;
+        QTextStream stream(&message);
+        stream << type << " cannot be meaningfully interpreted as a shape type by ConstructShape method";
+        GeneralException(message).raise();
+        return nullptr;
+    }
+}
+
+Shape* Data::ConstructText(QString& type, QList<qint32>& dimensions, QString& body, QFont& font)
+{
+    if(type == TEXT)
```

Comment:
Isn't this a bit redundant? To get to this function, the type *MUST BE* of type ``Text``

Reply:
Good point. I decided to put the exception handling at a higher level than this low-level helper function so the style doesn't look so strange